### PR TITLE
feat(components/table): extract xm-table directive

### PR DESCRIPTION
**Pull request by Kaskyi**
refactor table-widget.
etract xm-table directive and following components.
remove XmConfigService
transfer chips-control

### CODE REVIEW

File: packages/components/table/controllers/collections/xm-table-collection-controller-resolver.service.ts [hunk 1]
```diff
@@ -60,35 +57,48 @@ export class XmTableCollectionControllerResolver<T = unknown> {
...
     }
 
-    public async get(): Promise<IXmTableCollectionController<T>> {
-        const config = await firstValueFrom(this.configController.config$());
-        const collectionType = config.collection.type;
+    public factory(collectionConfig: XmTableCollectionControllerType): IXmTableCollectionController<T> {
+        const collectionType = collectionConfig.type;
 
+        let collectionController: IXmTableCollectionController<T> & XmDynamicWithConfig;
         switch (collectionType) {
```
Comment (denisogolubev) [inline packages/components/table/controllers/collections/xm-table-collection-controller-resolver.service.ts:68]: The previous implementation of the switch case was better. Less code, less duplication, easier to expand cases.
It's better to separate switch case logic to other method and in this method just call it. 
```ts
const collectionController: IXmTableCollectionController<T> & XmDynamicWithConfig = this.getCollectionController(collectionType);
```

File: packages/components/table/controllers/elastic/xm-table-filters-elastic-string-query.ts [hunk 2]
```diff
@@ -17,7 +18,11 @@ import {
     userFullnameElastic,
 } from './xm-table-filter-elastic';
 
-export const Xm_TABLE_FILTERS_ELASTIC_STRING_QUERY = {
+export interface XmTableFiltersElasticStringQuery {
```
Comment (denisogolubev) [inline packages/components/table/controllers/elastic/xm-table-filters-elastic-string-query.ts:21]: Why not in the model file?

File: packages/components/table/directives/xm-table-mat-paginator-adapter.directive.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    standalone: true,
+})
+export class XmTableMatPaginatorAdapterDirective implements AfterViewInit, OnDestroy {
+    constructor(
+        @Self() private matPaginator: MatPaginator,
+        private xmTableDirective: XmTableDirective,
+    ) {
+    }
+
+    public ngAfterViewInit(): void {
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table-mat-paginator-adapter.directive.ts:19]: It's not a good practice to expand code in Angular lifecycle hooks.
It's better to create separated method and call it in lifecycle hook.

Also, it's two subscriptions that can be handled via one subscription. Optimize this code, please.

File: packages/components/table/directives/xm-table-mat-sort-adapter.directive.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+export class XmTableMatSortAdapterDirective implements AfterViewInit, OnDestroy {
+
+    constructor(
+        @Self() private matSort: MatSort,
+        private xmTableDirective: XmTableDirective,
+    ) {
+    }
+
+    public ngAfterViewInit(): void {
+        this.xmTableDirective.context$
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table-mat-sort-adapter.directive.ts:21]: the same as previous comment

File: packages/components/table/directives/xm-table-selection.directive.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        XmTableSelectionService,
+    ],
+    standalone: true,
+})
+export class XmTableSelectionDirective {
+    @Input('xmTableSelectionConfig')
+    public config: XmTableConfig;
+
+    public get selectColumn(): XmTableSelectTableColumn {
+        const overrideConfig = this.config.columns.find(c => c.name == '_selectColumn') as XmTableSelectTableColumn || null;
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table-selection.directive.ts:25]: The magic string that repeated twice in the file `'_selectColumn'`. Use Enum or read-only variable for it.

File: packages/components/table/directives/xm-table.directive.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    ) {
+    }
+
+    private _config: XmTableConfig;
+
+    public get config(): XmTableConfig {
+        return this._config;
+    }
+
+    @Input('xmTableConfig')
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table.directive.ts:67]: It's not good practice to alias @Inputs
https://angular.io/guide/styleguide#avoid-aliasing-inputs-and-outputs

File: packages/components/table/directives/xm-table.directive.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    public set config(value: XmTableConfig) {
+        this._config = _.defaultsDeep({}, value, cloneDeep(XM_TABLE_CONFIG_DEFAULT));
+        this.columnsSettingStorageService.defaultStore(getDisplayedColumns(this._config));
+    }
+
+    public ngOnInit(): void {
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table.directive.ts:73]: It's not a good practice to expand code in Angular lifecycle hooks.
Separate this code into different methods and call these methods in ngOnInit.

File: packages/components/table/directives/xm-table.directive.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    public ngOnDestroy(): void {
+        takeUntilOnDestroyDestroy(this);
+    }
+
+    public updatePagination(): void {
+        const sortBy = this._config.columns.find((i) => i.name === this.sort.active)?.name;
+        const sortOrder = this.sort.direction;
+        const pageIndex = this.paginator.pageIndex;
```
Comment (denisogolubev) [inline packages/components/table/directives/xm-table.directive.ts:112]: could be simplified via de structurization:
```ts
const {pageIndex, pageSize, length} = this.paginator || {};
```

File: packages/components/table/cells/xm-table-edit-cell.component.ts [hunk 9]
```diff
@@ -42,13 +39,13 @@ export class XmTableEditCellComponent implements OnInit {
     constructor(
         @Inject(XM_DYNAMIC_TABLE_ROW) public row: any,
         @Inject(XM_DYNAMIC_TABLE_CELL) public cell: any,
-        private collectionControllerResolver: XmTableCollectionControllerResolver) {
+        private tableDirective: XmTableDirective) {
     }
 
-    public async ngOnInit(): Promise<void> {
-        const collection = await this.collectionControllerResolver.get();
+    public ngOnInit(): void {
```
Comment (denisogolubev) [inline packages/components/table/cells/xm-table-edit-cell.component.ts:45]: It's not a good practice to expand code in Angular lifecycle hooks.
Separate this code into different methods and call these methods in ngOnInit.

**Review by grankovsky — APPROVED** [review]

File: packages/components/table/controllers/collections/xm-table-collection-controller-resolver.service.ts (hunk 1 above)
Comment (Kaskyi) [reply packages/components/table/controllers/collections/xm-table-collection-controller-resolver.service.ts:68]: done

File: packages/components/table/controllers/elastic/xm-table-filters-elastic-string-query.ts (hunk 2 above)
Comment (Kaskyi) [reply packages/components/table/controllers/elastic/xm-table-filters-elastic-string-query.ts:21]: imho its not convinient

done

File: packages/components/table/directives/xm-table-mat-paginator-adapter.directive.ts (hunk 3 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table-mat-paginator-adapter.directive.ts:19]: done

simplisity over optimisation

File: packages/components/table/directives/xm-table-mat-sort-adapter.directive.ts (hunk 4 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table-mat-sort-adapter.directive.ts:21]: done

File: packages/components/table/directives/xm-table-selection.directive.ts (hunk 5 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table-selection.directive.ts:25]: done

File: packages/components/table/directives/xm-table.directive.ts (hunk 6 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table.directive.ts:67]: done

File: packages/components/table/directives/xm-table.directive.ts (hunk 7 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table.directive.ts:73]: done

File: packages/components/table/directives/xm-table.directive.ts (hunk 8 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table.directive.ts:112]: still good as it is

File: packages/components/table/cells/xm-table-edit-cell.component.ts (hunk 9 above)
Comment (Kaskyi) [reply packages/components/table/cells/xm-table-edit-cell.component.ts:45]: done

**Review by LevBD — APPROVED** [review]

File: packages/components/table/components/xm-table-loading-column.component.ts [hunk 10]
```diff
@@ -9,15 +9,15 @@ import { CdkCellDef, CdkColumnDef, CdkHeaderCellDef, CdkTable } from '@angular/c
...
+                [attr.colspan]="colspan">
                 <xm-loading-bar [visible]="isLoading"></xm-loading-bar>
             </th>
-            <td mat-cell *matCellDef [attr.colspan]="size"></td>
+            <td mat-cell *matCellDef [attr.colspan]="colspan"></td>
         </ng-container>
     `,
     styles: [`
-        th.loadingCell {
+        th.xm-table-loading-column-cell {
```
Comment (Bozonych) [inline packages/components/table/components/xm-table-loading-column.component.ts:20]: You can remove it, and use boostrap classes: p-0 border-0

File: packages/components/table/directives/xm-table.directive.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+export interface IXmTableContext {
+    collection: IXmTableCollectionState<unknown>,
+    settings: { displayedColumns: string [] };
+}
+
+function getDisplayedColumns(config: XmTableConfig): ColumnsSettingStorageItem[] {
+    const displayedColumns = config.columns;
+    return displayedColumns.map(i => ({
+        name: i.name || i.field,
```
Comment (Bozonych) [inline packages/components/table/directives/xm-table.directive.ts:28]: For better reading this code, make sense change i => column. 
Tthen the next line will be: `name: cloumn.name || column.field`

**Review by Bozonych — APPROVED** [review]

File: packages/components/table/directives/xm-table.directive.ts (hunk 11 above)
Comment (Kaskyi) [reply packages/components/table/directives/xm-table.directive.ts:28]: done

File: packages/components/table/components/xm-table-loading-column.component.ts (hunk 10 above)
Comment (Kaskyi) [reply packages/components/table/components/xm-table-loading-column.component.ts:20]: i do prefer to keep individual style over bootrap. this class verides material styles, and see that it mihgt be change in future.

**Pull request merged**